charms/kserve-controller/src/charm.py

```diff
@@ -362,105 +381,54 @@ def _generate_gateways_context(self) -> dict:
 
         return gateways_context
 
-    def gen_certs(self, namespace, service_name):
-        """Generate certificates."""
-        if Path("/run/cert.pem").exists():
-            log.info("Found existing cert.pem, not generating new cert.")
-            return
-
-        Path("/run/ssl.conf").write_text(
-            f"""[ req ]
-default_bits = 2048
-prompt = no
-default_md = sha256
-req_extensions = req_ext
-distinguished_name = dn
-[ dn ]
-C = GB
-ST = Canonical
-L = Canonical
-O = Canonical
-OU = Canonical
-CN = 127.0.0.1
-[ req_ext ]
-subjectAltName = @alt_names
-[ alt_names ]
-DNS.1 = {service_name}
-DNS.2 = {service_name}.{namespace}
-DNS.3 = {service_name}.{namespace}.svc
-DNS.4 = {service_name}.{namespace}.svc.cluster
-DNS.5 = {service_name}.{namespace}.svc.cluster.local
-DNS.6 = kserve-webhook-server-service
-DNS.7 = kserve-webhook-server-service.{namespace}
-DNS.8 = kserve-webhook-server-service.{namespace}.svc
-DNS.9 = kserve-webhook-server-service.{namespace}.svc.cluster
-DNS.10 = kserve-webhook-server-service.{namespace}.svc.cluster.local
-IP.1 = 127.0.0.1
-[ v3_ext ]
-authorityKeyIdentifier=keyid,issuer:always
-basicConstraints=CA:FALSE
-keyUsage=keyEncipherment,dataEncipherment,digitalSignature
-extendedKeyUsage=serverAuth,clientAuth
-subjectAltName=@alt_names"""
+    def _gen_certs_if_missing(self):
+        """Generate certificates if they don't already exist in _stored."""
+        log.info("Generating certificates if missing.")
+        cert_attributes = ["cert", "ca", "key"]
+        # Generate new certs if any cert attribute is missing
+        for cert_attribute in cert_attributes:
+            try:
+                getattr(self._stored, cert_attribute)
+            except AttributeError:
+                self._gen_certs()
+                return
+        log.info("Certificates already exist.")
+
+    def _gen_certs(self):
+        """Refresh the certificates, overwriting them if they already existed."""
```

Comment:
Is there a case when the certs are overwritten?

Reply:
existing cert attributes might be overwritten if any other attribute is missing, let me update that docstring

Reply:
Oh that makes sense, yeah